File: js/process_service.js
var FindService = require('./find_service');

var service = {
	findService: FindService(),

	run: function (input) {
		if (!input) throw new Error('Input invalid');
		this.input = input;
		return this.complexProcessing();
	},

	complexProcessing: function () {
		var self = this;
		var found = this.findService.run();

		return found.map(function (e) {
			return e * self.input;
		});
	}
}

module.exports = function () {
	return Object.create(service);
}

File: ruby/process_service.rb
require_relative "find_service"

class ProcessService

	attr_writer :find_service

	def run(input)
		@input = input
		complex_processing
	end

	def complex_processing
		items = find_service.run
		items.map do |i|
			i * @input
		end
	end

	def find_service
		@find_service ||= FindService.new
	end

end